pyro/ops/tensor_utils.py

```diff
@@ -395,7 +395,15 @@ def inverse_haar_transform(x):
 
 def cholesky(x):
     if x.size(-1) == 1:
+        x = x.clamp(min=torch.finfo(x.dtype).tiny)
         return x.sqrt()
+
+    # Add adaptive jitter.
+    x = x.clone()
```

Comment:
are you sure you need/want clones? fwiw i do [this](https://github.com/BasisResearch/millipede/blob/ecafe4194e216846600ff4f87dc02f65ce4428e2/millipede/util.py#L16) in millipede, which is similar to what's done in gpytorch

Reply:
Yes this `.clone()` is needed because we're mutating the matrix.

Thanks, I did try various gpytorch-style lazy tactics that avoid adding jitter until a failure has occurred.  I found that since the Cholesky error happens late in the filtering process, by that point the filter state had already been corrupted by nearly-singular matrices that just barely didn't trigger an error.  The best solution I've found so far is to add a tiny amount of noise to all matrices so that error doesn't build up during filtering.  Other solutions include using `svd` or `pinv` or `ldl_factor`, but they were more expensive.

Note the core piece of linear algebra is `.marginalize()` that is repeatedly called in the filter pass of `sequential_gaussian_filter_sample()`.  This has the nice property that if we add a little bit of jitter before Cholesky factorizing, the resulting `precision` becomes only better-conditioned.  Empirically this allows us to get away with much smaller jitter than would be required if we waited for an error to occur:
```python
P_aa = self.precision[..., a, a]
P_ba = self.precision[..., b, a]
P_bb = self.precision[..., b, b]
P_b = safe_cholesky(P_bb)                       # Note if we add a little jitter here...
P_a = triangular_solve(P_ba, P_b, upper=False)  # ...then this is smaller...
P_at = P_a.transpose(-1, -2)
precision = P_aa - matmul(P_at, P_a)            # ...so this is even better conditioned.
```

BTW it looks like you could speed up millipede by switching from `try: c = cholesky()` to the faster `c, info = cholesky_ex(); if not info.any(): return c`, which is used in gpytorch.  The only reason I'm not using `cholesky_ex()` here is that I found the decision-based version was too unstable.